// @fluentui/utilities/lib-amd/shadowDom/hooks/useMergeStylesHooks.js
define(["require", "exports", "react", "../contexts/MergeStylesDefaultContext"], function (require, exports, React, MergeStylesDefaultContext_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.useMergeStylesHooks = void 0;
    var useMergeStylesHooks = function () {
        var ctx = React.useContext(MergeStylesDefaultContext_1.MergeStylesDefaultContext);
        return {
            useAdoptedStylesheet: ctx.useAdoptedStylesheet,
            useAdoptedStylesheetEx: ctx.useAdoptedStylesheetEx,
            useShadowConfig: ctx.useShadowConfig,
            useMergeStylesShadowRootContext: ctx.useMergeStylesShadowRootContext,
            useHasMergeStylesShadowRootContext: ctx.useHasMergeStylesShadowRootContext,
            useMergeStylesRootStylesheets: ctx.useMergeStylesRootStylesheets,
            useWindow: ctx.useWindow,
            useStyled: ctx.useStyled,
        };
    };
    exports.useMergeStylesHooks = useMergeStylesHooks;
});
